### ISSUE TITLE
Syntax error in shader

### ISSUE THREAD

**Issue by kephale** [scenerygraphics/scenery#296]
```
[SciJava-483098e-Thread-0] INFO RAIVolume - Composing model for DelegatesRendering(RAIVolume)
[SciJava-483098e-Thread-0] INFO RAIVolume - Sizes are [[1000.0, 1000.0, 1000.0]]
[SciJava-483098e-Thread-0] INFO RAIVolume - Local scale is [[0.1, 0.1, 0.1]]
[SciJava-483098e-Thread-0] INFO Colormap - Using colormap viridis from stream
[SciJava-483098e-Thread-0] INFO Colormap - Read image from sun.net.www.protocol.jar.JarURLConnection$JarURLInputStream@5be0ea29 with 20480 bytes, size=256x20
[SciJava-483098e-Thread-0] INFO VolumeManager - Adding DelegatesRendering(RAIVolume) to OOC nodes
[SciJava-483098e-Thread-0] INFO VolumeManager - Updating state for 3 BDV nodes
[SciJava-483098e-Thread-0] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-0] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-0] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-1] INFO VolumeManager - Updating state for 3 BDV nodes
[SciJava-483098e-Thread-0] INFO VolumeManager - (0, 2) -> ooc:0 reg:3
[SciJava-483098e-Thread-1] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-1] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-1] INFO VolumeManager - ST is 3d-affine: (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0)
[SciJava-483098e-Thread-1] INFO VolumeManager - (0, 2) -> ooc:0 reg:3
[SciJava-483098e-Thread-1] INFO VolumeManager - Using program for 0 out-of-core volumes and 3 regular volumes
[SciJava-483098e-Thread-1] INFO VolumeManager - We have 3 shaders ready
[SciJava-483098e-Thread-1] INFO VolumeManager - Updating effective shader program to tpietzsch.example2.MultiVolumeShaderMip@c6c8c34
[SciJava-483098e-Thread-1] INFO SceneryStackManager - U8 texture does not exist, creating new one (tpietzsch.example2.VolumeTextureU8@31e6e98f)
[SciJava-483098e-Thread-1] INFO SceneryStackManager - U8 texture does not exist, creating new one (tpietzsch.example2.VolumeTextureU8@24c1349)
[SciJava-483098e-Thread-1] INFO SceneryStackManager - U8 texture does not exist, creating new one (tpietzsch.example2.VolumeTextureU8@2ed1b830)
[SciJava-483098e-Thread-1] ERROR SceneryContext - Binding for graphics.scenery.volumes.bdv.VolumeManager$SimpleTexture2D@2babda28 not found, despite trying deferred binding. (binding=BindingState(binding=9, uniformName=null, reallocate=false)/sampler=null)
[SciJava-483098e-Thread-1] ERROR SceneryContext - Binding for tpietzsch.example2.VolumeTextureU8@2716e55 not found, despite trying deferred binding. (binding=null/sampler=null)
[SciJava-483098e-Thread-1] ERROR SceneryContext - Binding for graphics.scenery.volumes.bdv.VolumeManager$SimpleTexture2D@46b49cf2 not found, despite trying deferred binding. (binding=BindingState(binding=10, uniformName=null, reallocate=false)/sampler=null)
[SciJava-483098e-Thread-0] INFO VolumeManager - Using program for 0 out-of-core volumes and 3 regular volumes
[SciJava-483098e-Thread-0] INFO VolumeManager - We have 4 shaders ready
[SciJava-483098e-Thread-0] INFO VolumeManager - Updating effective shader program to tpietzsch.example2.MultiVolumeShaderMip@6f8035ca
[SciJava-483098e-Thread-0] INFO Colormap - Using ImageJ colormap net.imglib2.display.ColorTable8@65786754 with size 256x16
[SciJava-483098e-Thread-1] ERROR VulkanRenderer - Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.lang.RuntimeException: Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] ERROR VulkanRenderer - Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.lang.RuntimeException: Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] ERROR VulkanRenderer - Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.lang.RuntimeException: Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.io.FileNotFoundException: fromBuffer:transferFunction_x_34_x_ (No such file or directory)
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.io.FileNotFoundException: fromBuffer:transferFunction_x_48_x_ (No such file or directory)
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.io.FileNotFoundException: fromBuffer:transferFunction_x_13_x_ (No such file or directory)
[SciJava-483098e-Thread-1] ERROR VulkanRenderer - Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanRenderer - Could not load texture for VolumeManager: java.lang.RuntimeException: Call to allocate image staging memory of size 1040449536 failed: A device memory allocation has failed.
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_35_x_ not found for pass Scene
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_49_x_ not found for pass Scene
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_35_x_ not found for pass DeferredLighting
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_49_x_ not found for pass DeferredLighting
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_35_x_ not found for pass ForwardShading
[SciJava-483098e-Thread-1] WARN VulkanObjectState - graphics.scenery.backends.vulkan.VulkanObjectState@3635a0d2: DSL for colorMap_x_49_x_ not found for pass ForwardShading
[SciJava-483098e-Thread-1] ERROR VolumeShaderFactory - Error in shader compilation of  for VolumeShaderFactory: ERROR: 0:129: '' :  syntax error, unexpected EQUAL, expecting COMMA or SEMICOLON
ERROR: 1 compilation errors.  No code generated.


[SciJava-483098e-Thread-1] ERROR VolumeShaderFactory - Shader code was: 
1	:  // preprocessor statements
2	:  
3	:  // VolumeShaderFactory: inputs
4	:  
5	:  // VolumeShaderFactory: outputs
6	:  layout (location = 0) out vec4 FragColor;
7	:  // VolumeShaderFactory: uniforms
8	:  // autogenerated ShaderProperties
9	:  layout(set = 2, binding = 0) uniform ShaderProperties {
10	:  	float fwnw;
11	:  	float nw;
12	:  	float xf;
13	:  	mat4 transform;
14	:  	vec2 dsp;
15	:  	vec2 viewportSize;
16	:  	vec3 blockSize;
17	:  	vec3 cachePadOffset;
18	:  	vec3 cacheSize; // TODO: get from texture!?
19	:  	vec3 paddedBlockSize;
20	:  };
21	:  
22	:  // VolumeShaderFactory: samplers
23	:  layout(set = 3, binding = 0) uniform sampler2D InputZBuffer;
24	:  layout(set = 4, binding = 0) uniform sampler3D volumeCache;
25	:  // VolumeShaderFactory: main code 
26	:  
27	:  float tw( float zd )
28	:  {
29	:  	return ( xf * zd ) / ( 2 * xf * zd - xf - zd + 1 );
30	:  }
31	:  
32	:  
33	:  float getMaxDepth( vec2 uv )
34	:  {
35	:  	return tw( texture( InputZBuffer, ( uv + 1 ) / 2 ).x );
36	:  	//#ifndef OPENGL
37	:  	//    float currentSceneDepth = texture(InputZBuffer, uv).r;
38	:  	//#else
39	:  	//    float currentSceneDepth = texture(InputZBuffer, uv).r * 2.0 - 1.0;
40	:  	//#endif
41	:  	//	return tw(currentSceneDepth);
42	:  }
43	:  
44	:  
45	:  // -- comes from CacheSpec -----
46	:  
47	:  // -- comes from TextureCache --
48	:  
49	:  layout(location = 0) in VertexData {
50	:  	vec2 textureCoord;
51	:  	mat4 inverseProjection;
52	:  	mat4 inverseView;
53	:  } Vertex;
54	:  
55	:  layout(set = 0, binding = 0) uniform VRParameters {
56	:  	mat4 projectionMatrices[2];
57	:  	mat4 inverseProjectionMatrices[2];
58	:  	mat4 headShift;
59	:  	float IPD;
60	:  	int stereoEnabled;
61	:  } vrParameters;
62	:  
63	:  const int MAX_NUM_LIGHTS = 1024;
64	:  
65	:  layout(set = 1, binding = 0) uniform LightParameters {
66	:  	mat4 ViewMatrices[2];
67	:  	mat4 InverseViewMatrices[2];
68	:  	mat4 ProjectionMatrix;
69	:  	mat4 InverseProjectionMatrix;
70	:  	vec3 CamPosition;
71	:  };
72	:  
73	:  layout(push_constant) uniform currentEye_t {
74	:  	int eye;
75	:  } currentEye;
76	:  
77	:  // intersect ray with a box
78	:  // http://www.siggraph.org/education/materials/HyperGraph/raytrace/rtinter3.htm
79	:  void intersectBox( vec3 r_o, vec3 r_d, vec3 boxmin, vec3 boxmax, out float tnear, out float tfar )
80	:  {
81	:  	// compute intersection of ray with all six bbox planes
82	:  	vec3 invR = 1 / r_d;
83	:  	vec3 tbot = invR * ( boxmin - r_o );
84	:  	vec3 ttop = invR * ( boxmax - r_o );
85	:  
86	:  	// re-order intersections to find smallest and largest on each axis
87	:  	vec3 tmin = min(ttop, tbot);
88	:  	vec3 tmax = max(ttop, tbot);
89	:  
90	:  	// find the largest tmin and the smallest tmax
91	:  	tnear = max( max( tmin.x, tmin.y ), max( tmin.x, tmin.z ) );
92	:  	tfar = min( min( tmax.x, tmax.y ), min( tmax.x, tmax.z ) );
93	:  }
94	:  
95	:  // ---------------------
96	:  
97	:  
98	:  
99	:  
100	:  // ---------------------
101	:  
102	:  void main()
103	:  {
104	:  	mat4 ipv = Vertex.inverseView * Vertex.inverseProjection;
105	:  	// frag coord in NDC
106	:  	// TODO: Re-introduce dithering
107	:  	//	vec2 fragCoord = (vrParameters.stereoEnabled ^ 1) * gl_FragCoord.xy + vrParameters.stereoEnabled * vec2((gl_FragCoord.x/2.0 + currentEye.eye * gl_FragCoord.x/2.0), gl_FragCoord.y);
108	:  	//	vec2 viewportSizeActual = (vrParameters.stereoEnabled ^ 1) * viewportSize + vrParameters.stereoEnabled * vec2(viewportSize.x/2.0, viewportSize.y);
109	:  	//	vec2 uv = 2 * ( gl_FragCoord.xy + dsp ) / viewportSizeActual - 1;
110	:  	vec2 uv = Vertex.textureCoord * 2.0 - vec2(1.0);
111	:  	vec2 depthUV = (vrParameters.stereoEnabled ^ 1) * Vertex.textureCoord + vrParameters.stereoEnabled * vec2((Vertex.textureCoord.x/2.0 + currentEye.eye * 0.5), Vertex.textureCoord.y);
112	:  	depthUV = depthUV * 2.0 - vec2(1.0);
113	:  
114	:  	// NDC of frag on near and far plane
115	:  	vec4 front = vec4( uv, -1, 1 );
116	:  	vec4 back = vec4( uv, 1, 1 );
117	:  
118	:  	// calculate eye ray in world space
119	:  	vec4 wfront = ipv * front;
120	:  	wfront *= 1 / wfront.w;
121	:  	vec4 wback = ipv * back;
122	:  	wback *= 1 / wback.w;
123	:  
124	:  	// -- bounding box intersection for all volumes ----------
125	:  	float tnear = 1, tfar = 0, tmax = getMaxDepth( depthUV );
126	:  	float n, f;
127	:  
128	:  	// 
129	:  	bool  = false;
130	:  	intersectBoundingBox_x_105_x_( wfront, wback, n, f );
131	:  	f = min( tmax, f );
132	:  	if ( n < f )
133	:  	{
134	:  		tnear = min( tnear, max( 0, n ) );
135	:  		tfar = max( tfar, f );
136	:   = true;
137	:  	}
138	:  	// 
139	:  
140	:  	// -------------------------------------------------------
141	:  
142	:  
143	:  	if ( tnear < tfar )
144	:  	{
145	:  		vec4 fb = wback - wfront;
146	:  		int numSteps =
147	:  		( fwnw > 0.00001 )
148	:  		? int ( log( ( tfar * fwnw + nw ) / ( tnear * fwnw + nw ) ) / log ( 1 + fwnw ) )
149	:  		: int ( trunc( ( tfar - tnear ) / nw + 1 ) );
150	:  
151	:  		float step = tnear;
152	:  		vec4 v = vec4( 0 );
153	:  		for ( int i = 0; i < numSteps; ++i, step += nw + step * fwnw )
154	:  		{
155	:  			vec4 wpos = mix( wfront, wback, step );
156	:  
157	:  
158	:  
159	:  			/*
160	:  			inserts something like the following (keys: ,blockTexture,convert)
161	:  
162	:  			if ()
163	:  			{
164	:  				float x = blockTexture(wpos, volumeCache, cacheSize, blockSize, paddedBlockSize, cachePadOffset);
165	:  				v = max(v, convert(x));
166	:  			}
167	:  			*/
168	:  		}
169	:  		FragColor = v;
170	:  	}
171	:  	else
172	:  	FragColor = vec4( 0, 0, 0, 0 );
173	:  }
174	:  
[SciJava-483098e-Thread-1] ERROR VolumeShaderFactory - Error in shader linking of  for VolumeShaderFactory: 
[SciJava-483098e-Thread-0] INFO RAIVolume - Composing model for DelegatesRendering(RAIVolume)
[SciJava-483098e-Thread-0] INFO RAIVolume - Sizes are [[1000.0, 1000.0, 1000.0]]
[SciJava-483098e-Thread-0] INFO RAIVolume - Local scale is [[0.1, 0.1, 0.1]]
```

**Comment by skalarproduktraum** [comment]
Do you have some details, e.g. what you did before this happened?

**Comment by kephale** [comment]
Just opening 3 in-core volumes (they might have exceeded memory capacity)

**Comment by skalarproduktraum** [comment]
So I think what happened is that the volumes were not initialised properly, because out of memory, which lead to the creation of a broken shader. This is of course not good, and should be prevented. I will investigate.

**Comment by skalarproduktraum** [comment]
This should be fixed with the merging of #292. Please reopen if it occurs nevertheless.
